### PR TITLE
Add bidirectional converter specs to test namespaces

### PR DESCRIPTION
**Pull request by MrSerth**
Previously, we might have duplicated namespaces when converting the same string back and forth. With the latest change in openHPI/dachsfisch#52, we're avoiding this error and are now adding respective test coverage for this phenomena.

### CODE REVIEW

File: spec/support/expectations/equal_namespace_definitions.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    @actual = actual_xml.to_xml
+
+    return false if expected_xml.errors.length.positive? || actual_xml.errors.length.positive?
+    return false unless EquivalentXml.equivalent?(expected_xml, actual_xml)
+
+    return compare_namespaces(actual_xml.children.first, expected_xml.children.first)
+  end
+
+  failure_message do |actual|
+    "#{@actual || actual} is not equal to \n#{@expected || expected}."
```
Comment (kkoehn) [inline spec/support/expectations/equal_namespace_definitions.rb:23]: I think it would be nice to have a more accurate error message (ie "does not have equal namespace as")

File: spec/support/expectations/equal_namespace_definitions.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  match do |actual|
+    return false unless actual.is_a?(String) && expected.is_a?(String)
+
+    expected_xml = parse_fragment(expected)
+    @expected = expected_xml.to_xml
+    actual_xml = parse_fragment(actual)
+    @actual = actual_xml.to_xml
+
+    return false if expected_xml.errors.length.positive? || actual_xml.errors.length.positive?
+    return false unless EquivalentXml.equivalent?(expected_xml, actual_xml)
```
Comment (kkoehn) [inline spec/support/expectations/equal_namespace_definitions.rb:17]: Is it necessary to check for XML-equality as well? It feels out of scope for namespace comparison.

File: spec/custom_matchers/equal_namespace_definitions_spec.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+      it 'provides a useful error message' do
+        expect { expect(xml).to be_an_equal_json_as xml2 }
+          .to raise_error(RSpec::Expectations::ExpectationNotMetError, /not equal to/)
+      end
+    end
+
+    context 'with similar namespace definitions' do
+      let(:xml2) do
+        <<~XML
```
[the remark this answers is not in the document]
Comment (kkoehn) [reply spec/custom_matchers/equal_namespace_definitions_spec.rb:85]: If we decide to remove the equality check, I would suggest changing the texts of the nodes here, otherwise I would reuse `xml`

**Review by kkoehn — COMMENTED** [review]
Looks great.
I added a few remarks with questions and suggestions

File: spec/custom_matchers/equal_namespace_definitions_spec.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (MrSerth) [reply spec/custom_matchers/equal_namespace_definitions_spec.rb:85]: Actually, this check was intended to have the same namespaces in two different orders. I missed that, but pushed a change now.

File: spec/support/expectations/equal_namespace_definitions.rb (hunk 2 above)
Comment (MrSerth) [reply spec/support/expectations/equal_namespace_definitions.rb:17]: Not necessarily. However, my implementation explicitly requires XML-equality. This is caused by the way I extract the namespace definitions and how I compare them: Nokogiri is very smart in handling the namespaces and performs a deduplication of namespaces while reading. Hence, I need to use the `namespace_definitions` method on each node to get the namespaces explicitly defined there. As a consequence, I need to traverse both node trees.

With the assumption that the trees are equal, it is enough to traverse one tree and perform the similar operation on the other tree. Otherwise, I would need to traverse both trees independently from each other and then extract the definitions at each node and level. When one of the trees is not as high as the other, I need to check whether the there are further namespaces definitions in this additional height.

This checking would make the matcher much more complex, so that I refrained from it (since it is not used either). As a consequence, I need to check for XML-quality.

Comment (kkoehn) [reply spec/support/expectations/equal_namespace_definitions.rb:17]: Makes sense, thanks for the explanation

**Review by kkoehn — APPROVED** [review]
